commit 8be78d3df47a43cfd8dde047026e7f238030e253
Author: Loren Merritt <pengvado@akuvian.org>
Date:   Mon Jun 17 11:27:09 2013 -0700

    Fix build with PIC on some systems

diff --git a/common/x86/cabac-a.asm b/common/x86/cabac-a.asm
index 74dbf93..e2e0d5e 100644
--- a/common/x86/cabac-a.asm
+++ b/common/x86/cabac-a.asm
@@ -62,6 +62,9 @@ cextern coeff_last64_sse2
 cextern coeff_last64_sse2_lzcnt
 cextern coeff_last64_avx2_lzcnt
 
+%ifdef PIC
+SECTION .data
+%endif
 coeff_last_sse2:       COEFF_LAST_TABLE       mmx2,       sse2,       sse2, 16, 15, 16, 4, 15, 64, 16, 15, 16, 64, 16, 15, 16, 64
 coeff_last_sse2_lzcnt: COEFF_LAST_TABLE mmx2_lzcnt, sse2_lzcnt, sse2_lzcnt, 16, 15, 16, 4, 15, 64, 16, 15, 16, 64, 16, 15, 16, 64
 coeff_last_avx2_lzcnt: COEFF_LAST_TABLE mmx2_lzcnt, avx2_lzcnt, sse2_lzcnt, 16, 15, 16, 4, 15, 64, 16, 15, 16, 64, 16, 15, 16, 64

