diff -uNr gpsdrive-2.11.orig/src/mapnik.cpp gpsdrive-2.11/src/mapnik.cpp
--- gpsdrive-2.11.orig/src/mapnik.cpp	2013-02-23 16:01:42.335966543 +0100
+++ gpsdrive-2.11/src/mapnik.cpp	2013-02-23 16:09:35.756593480 +0100
@@ -30,6 +30,9 @@
 #include "config.h"
 #include "gpsdrive_config.h"
 
+#if MAPNIK_VERSION >= 200100
+#include <mapnik/graphics.hpp>
+#endif
 
 #if MAPNIK_VERSION < 200000
 using mapnik::Image32;
