--- configure	2009-08-04 01:46:33.932416922 +0200
+++ configure	2009-08-04 01:46:53.304202942 +0200
@@ -871,7 +871,7 @@
 our $failed = 0;
 
 $config{CERTGEN} ||= 'y';
-yesno('CERTGEN',"Would you like generate SSL certificates now?") if $config{USE_GNUTLS} eq "y" || $config{USE_OPENSSL} eq "y";
+yesno('CERTGEN',"Would you like generate SSL certificates now?") if ($interactive && ($config{USE_GNUTLS} eq "y" || $config{USE_OPENSSL} eq "y"));
 
 if ($config{USE_GNUTLS} eq "y") {
 	unless (-r "src/modules/m_ssl_gnutls.cpp") {
